python-threatexchange/threatexchange/threat_updates.py

```diff
@@ -0,0 +1,375 @@
+    """
+
+    def __init__(
+        self,
+        privacy_group: int,
+        start: int = 0,
+        end: t.Optional[int] = None,
+    ) -> None:
+        self.privacy_group = privacy_group
+        self.updates = []
+        self.current = start
+        self.start = start
+        self.end = end
+
+        self._cursor = None
+
+    @property
+    def done(self) -> bool:
+        """Has this delta fetched its entire assigned range?"""
+        return self.end and self.end <= self.current
+
+    def __iter__(self):
+        return iter(self.updates)
+
+    def merge(self, delta: "ThreatUpdatesDelta") -> "ThreatUpdatesDelta":
+        if not self.done or self.end != delta.start:
+            raise ValueError("unchecked merge!")
+        self.updates.extend(delta.updates)
+        self.current = delta.current
+        self.end = delta.end
+        self.tries = delta.tries
+
+    def one_fetch(self, api: ThreatExchangeAPI):
+        """
+        Do a single fetch from ThreatExchange and store the results.
+
+        One fetch only, please.
+        """
+        if self.done:
+            return
+        now = time.time()
+        if not self._cursor:
+            self._cursor = api.get_threat_updates(
+                self.privacy_group,
+                page_size=500,
+                start_time=self.start,
+                stop_time=self.end,
+                fields=ThreatUpdateJSON.te_threat_updates_fields(),
+                decode_fn=ThreatUpdateJSON,
+            )
+        for update in self._cursor.next():
+            self.updates.append(ThreatUpdateJSON(update.raw_json))
+            # Is supposed to be strictly increasing
+            self.current = max(update.time, self.current)
+        if self._cursor.done:
+            if not self.end:
+                self.end = int(now)
+            self.current = self.end
+
+        return self._cursor.data
+
+    def split(
+        self, n: int
+    ) -> t.Tuple["ThreatUpdatesDelta", t.List["ThreatUpdatesDelta"]]:
+        """Split this delta into n deltas of roughly even size"""
+        tar = self.end or time.time()
+        diff = (tar - self.current_time) // (n + 1)
+        if diff <= 0:
+            return self, []
+        end = self.end
+        prev = self
+        new_deltas = []
+        for i in range(n - 1):
+            prev.end = prev.start + diff
+            new_deltas.append(
+                ThreatUpdatesDelta(
+                    self.privacy_group,
+                    prev.end,
+                )
+            )
+        prev.end = end
+
+        return self, new_deltas
+
+    def incremental_sync_from_threatexchange(
+        self,
+        api: ThreatExchangeAPI,
+        *,
+        limit: int = 0,
+        progress_fn=lambda x: None,
+    ) -> None:
+        """
+        Fetch from threat_updates to get a more up-to-date copy of the data.
+        """
+
+        # TODO actually implement fancy threading logic
```

Comment:
ignorable: flagging that more error handingly is probably needed if you thread this because we try to save on error (might we worth a reminder in the comment).

Reply:
Will punt on this until the threading is written, but agree